### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by mikeynap**
Framework for authorization.
auth has a Tokenizer (JWT) and an authorizer (will be LDAP or DB etc).

Authorizer takes username/password and validates.
Tokenizer creates a token.

Added /v1/login endpoint:
Given username, password returns Token.

This Token is used in all requests. There is middleware around all the calls that requires the token set.

Currently errthing is set to the  null versions, so login() is not required and the endpoints are unprotected.

### CODE REVIEW

File: api/v1/client/token.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * OpenAPI spec version: 1.0
+ *
+ * Generated by: https://github.com/swagger-api/swagger-codegen.git
+ */
+
+package apiclient
+
+type Token struct {
+	Token   string `json:"token,omitempty"`
+	Expires int64  `json:"duration,omitempty"`
```
Comment (meganokeefe) [inline api/v1/client/token.go]: how is `Expires` expressed? seconds, milliseconds?

File: api/v1/gen/v1.yaml [hunk 2]
```diff
@@ -56,6 +74,18 @@ parameters:
...
 # Describe your paths here
 paths:
+  /login:
+    get:
+      security:
+        - Basic: []
+      description: generate token using basic auth.
+      operationId: loginWithStoredUsernameAndPassword
+      responses:
+        200:
```
Comment (meganokeefe) [inline api/v1/gen/v1.yaml]: shouldn't it be possible to have a non-200 response here?

File: api/v1/server/server.go [hunk 3]
```diff
@@ -180,24 +188,182 @@ func (s *Server) Stop() {
...
+	if err != nil {
+		log.Error(err)
+		w.Header().Set("WWW-Authenticate", `Basic realm="Enter Login for API Token"`)
+		http.Error(w, "Bad Login", http.StatusUnauthorized)
+		return
+	}
+	apiTok := convert.FromAuthToken(tok)
+	w.WriteHeader(http.StatusOK)
+	if err = json.NewEncoder(w).Encode(apiTok); err != nil {
+		log.Errorf("Unable to Encode Token: %v", err)
```
Comment (meganokeefe) [inline api/v1/server/server.go]: do we want to return an http error here? bad request?

File: api/v1/server/server.go [hunk 4]
```diff
@@ -180,24 +188,182 @@ func (s *Server) Stop() {
...
+	if users, err = s.auth.GetUsernames(nil); err != nil {
+		if err == user.ErrNotAllowed {
+			http.Error(w, err.Error(), http.StatusUnauthorized)
+			return
+		}
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if users == nil {
+		users = make([]string, 0)
```
Comment (meganokeefe) [inline api/v1/server/server.go]: what is this for?

File: api/v1/server/server.go [hunk 5]
```diff
@@ -180,24 +188,182 @@ func (s *Server) Stop() {
...
+
 	modObj := convert.FromAPIPlugin(apiObj)
 	var theID string
 	var err error
 	if theID, err = s.db.InsertPlugin(modObj); err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
-	w.Header().Add("Location", fmt.Sprintf("/v1/services/%v", theID))
+	w.Header().Add("Location", fmt.Sprintf("/v1/plugins/%v", theID))
```
Comment (meganokeefe) [inline api/v1/server/server.go]: good catch

File: auth/authorizer/db.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
+package authorizer
```
Comment (meganokeefe) [inline auth/authorizer/db.go]: unit tests for these functions plz?

File: auth/authorizer/db.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+	if group != "" {
+		su.AuthGroup = group
+	}
+
+	return a.sdb.UpdateSecret(su)
+}
+
+func (a *DBAuth) GetUser(un string) (user.User, error) {
+	u, err := a.sdb.GetSecret(newSecretUser(un))
```
Comment (meganokeefe) [inline auth/authorizer/db.go]: is this just-in-time struct creation for DB insertion idiomatic?
You're converting to-and-from here. 

Also, why is Secret an interface in `database.go`, vs a struct in `model.Go`?

File: auth/authorizer/db.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+	var suser *secretUser
+	var ok bool
+	if suser, ok = u.(*secretUser); !ok || suser == nil || suser.Id == "" {
+		return nil, ErrBadDatabaseResults
+	}
+	return suser.AuthUser, nil
+}
+
+func (a *DBAuth) GetUsernames(filters []string) ([]string, error) {
```
Comment (meganokeefe) [inline auth/authorizer/db.go]: are you documenting what filters should look like anywhere? 

is it a list like `["x=x_val", "y=y_val"]`?

File: auth/authorizer/db.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	filtMap := map[string]string{}
+	for _, f := range filters {
+		parts := strings.Split(f, "=")
+		if len(parts) != 2 {
```
Comment (meganokeefe) [inline auth/authorizer/db.go]: you're not rejecting invalid filters? do you want to `log.Warn` or anything here?

File: auth/authorizer/db.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	users, err := a.sdb.QuerySecrets(newSecretUser(""), filtMap)
+	if err != nil {
+		return nil, err
+	}
+	return users, nil
+}
+
+func (a *DBAuth) Login(un, pw string) (user.User, error) {
+	u, err := a.sdb.GetSecret(newSecretUser(un))
```
Comment (meganokeefe) [inline auth/authorizer/db.go]: see comment about typing above

File: auth/tokenizer/jwt.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
+package tokenizer
```
Comment (meganokeefe) [inline auth/tokenizer/jwt.go]: unit tests please

File: database/couchbase.go [hunk 12]
```diff
@@ -222,7 +222,7 @@ func (cb *Couchbase) IdFromName(n string, oType string) string {
 	if n == "" {
 		return ""
 	}
-	rows, err := cb.executeQuery("SELECT Id FROM `"+cb.cfg.BucketName+"` WHERE Name=$name AND Metadata.Type=$type", map[string]interface{}{
+	rows, err := cb.executeQuery("SELECT Id FROM `"+cb.cfg.BucketName+"` WHERE (Name=$name OR Metadata.Name=$name) AND Metadata.Type=$type", map[string]interface{}{
```
Comment (meganokeefe) [inline database/couchbase.go]: why this change?

File: database/couchbase.go [hunk 13]
```diff
@@ -609,12 +612,8 @@ func isClean(s string) bool {
 	return reg.MatchString(s)
 }
 
-func (cb Couchbase) makeQuery(obj model.Metadataer, idOnly bool) (string, map[string]interface{}, error) {
+func (cb Couchbase) makeQuery(obj model.Metadataer, selector string, filters map[string]string) (string, map[string]interface{}, error) {
```
Comment (meganokeefe) [inline database/couchbase.go]: need to add Secret query info to the switch in `makeQuery`?

File: debug.yaml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+---
```
Comment (meganokeefe) [inline debug.yaml]: should this config be merged into master?

lol `jitsNOTnits`

File: api/v1/server/server.go (hunk 3 above)
Comment (mikeynap) [inline api/v1/server/server.go]: Unfortunately you cannot -- Once you write to (w) you cannot write anymore headers. 

This is how we do it in the whole project, if we wanna try something else we'll do it in a different PR.

File: api/v1/client/token.go (hunk 1 above)
Comment (mikeynap) [inline api/v1/client/token.go]: It's a unix date.

File: api/v1/server/server.go (hunk 4 above)
Comment (mikeynap) [inline api/v1/server/server.go]: So that JSON.Encode writes [] and not an empty body.

File: auth/authorizer/db.go (hunk 7 above)
Comment (mikeynap) [inline auth/authorizer/db.go]: We'll chat about this --

File: auth/authorizer/db.go (hunk 8 above)
Comment (mikeynap) [inline auth/authorizer/db.go]: yes, I'll document.

File: debug.yaml (hunk 14 above)
Comment (mikeynap) [inline debug.yaml]: No it shouldn't -- I'll git rm.
